File: codes/kotlin/009.operator_optional_chaining.kt
@Suppress("UNUSED_VARIABLE", "UNCHECKED_CAST")
// to compile and run the code:
// kotlinc operator_optional_chaining.kt -include-runtime -d operator_optional_chaining.jar && kotlin operator_optional_chaining.jar

fun main() {
    println("\n// JavaScript-like Optional Chaining Operator (?.) in Kotlin")

    fun prettyJsonStringify(anything: Any? = null, indent: String = "    "): String {
        var indentLevel = 0
        fun prettyJsonStringifyInnerFunction(anything: Any? = null, indent: String = "    "): String {
            if (anything == null) return "undefined"
            if (anything == "null") return "null"
            if (anything == "undefined") return "undefined"
            if (anything is String) return "\"${anything}\""
            if (anything is Number || anything is Boolean) return "${anything}"
            if (anything is MutableList<*>) {
                indentLevel += 1
                var result = "[\n${indent.repeat(indentLevel)}"
                for ((arrayItemIndex, arrayItem) in anything.withIndex()) {
                    result += prettyJsonStringifyInnerFunction(arrayItem, indent)
                    if ((arrayItemIndex + 1) != anything.size) {
                        result += ",\n${indent.repeat(indentLevel)}"
                    }
                }
                indentLevel -= 1
                result += "\n${indent.repeat(indentLevel)}]"
                return result
            }
            if (anything is MutableMap<*, *>) {
                indentLevel += 1
                var result = "{\n${indent.repeat(indentLevel)}"
                anything.entries.forEachIndexed { entryIndex, entryItem ->
                    val objectKey = entryItem.key
                    val objectValue = entryItem.value
                    result += "\"${objectKey}\": ${prettyJsonStringifyInnerFunction(objectValue, indent)}"
                    if ((entryIndex + 1) != anything.entries.size) {
                        result += ",\n${indent.repeat(indentLevel)}"
                    }
                }
                indentLevel -= 1
                result += "\n${indent.repeat(indentLevel)}}"
                return result
            }
            return "undefined"
        }
        return prettyJsonStringifyInnerFunction(anything, indent)
    }

    // There's no JavaScript-like Optional Chaining Operator (?.) in Kotlin.
    // But, we can use Kotlin try-catch block to mimic it in Kotlin.

    val JSON_OBJECT = mutableMapOf<String, Any?>(
        "foo" to mutableMapOf<String, Any?>(
            "bar" to "baz",
        ),
        "fruits" to mutableListOf<Any?>("apple", "mango", "banana")
    )
    println("JSON_OBJECT: ${prettyJsonStringify(JSON_OBJECT)}")

    println("JSON_OBJECT?.foo?.bar: ${try { (JSON_OBJECT["foo"] as MutableMap<String, Any?>)["bar"] } catch (err: Exception) { null }}")
    // JSON_OBJECT?.foo?.bar: baz

    println("JSON_OBJECT?.foo?.baz: ${try { (JSON_OBJECT["foo"] as MutableMap<String, Any?>)["baz"] } catch (err: Exception) { null }}")
    // JSON_OBJECT?.foo?.baz: null // undefined and null is null in Kotlin

    println("JSON_OBJECT?.foo?.bar: ${prettyJsonStringify(try { (JSON_OBJECT["foo"] as MutableMap<String, Any?>)["bar"] } catch (err: Exception) { null })}")
    // JSON_OBJECT?.foo?.bar: "baz"

    println("JSON_OBJECT?.foo?.baz: ${prettyJsonStringify(try { (JSON_OBJECT["foo"] as MutableMap<String, Any?>)["baz"] } catch (err: Exception) { null })}")
    // JSON_OBJECT?.foo?.baz: undefined

    println("JSON_OBJECT?.fruits?.[2]: ${try { (JSON_OBJECT["fruits"] as MutableList<Any?>)[2] } catch (err: Exception) { null }}")
    // JSON_OBJECT?.fruits?.[2]: banana

    println("JSON_OBJECT?.fruits?.[5]: ${try { (JSON_OBJECT["fruits"] as MutableList<Any?>)[5] } catch (err: Exception) { null }}")
    // JSON_OBJECT?.fruits?.[5]: null // undefined and null is null in Kotlin

    println("JSON_OBJECT?.fruits?.[2]: ${prettyJsonStringify(try { (JSON_OBJECT["fruits"] as MutableList<Any?>)[2] } catch (err: Exception) { null })}")
    // JSON_OBJECT?.fruits?.[2]: "banana"

    println("JSON_OBJECT?.fruits?.[5]: ${prettyJsonStringify(try { (JSON_OBJECT["fruits"] as MutableList<Any?>)[5] } catch (err: Exception) { null })}")
    // JSON_OBJECT?.fruits?.[5]: undefined
}

File: codes/javascript/src/009_operator_optional_chaining.js
/* eslint-disable dot-notation */

const jsType = { "Null": "Null", "Boolean": "Boolean", "String": "String", "Numeric": "Numeric", "Object": "Object", "Array": "Array", "Function": "Function" };

const isNull = (anything) => (((Object.prototype.toString.call(anything) === "[object Null]") && (anything === null)) || ((Object.prototype.toString.call(anything) === "[object Undefined]") && (anything === undefined)));

const isBoolean = (anything) => ((Object.prototype.toString.call(anything) === "[object Boolean]") && ((anything === true) || (anything === false)));

const isString = (anything) => (Object.prototype.toString.call(anything) === "[object String]");

const isNumeric = (anything) => ((Object.prototype.toString.call(anything) === "[object Number]") && (Number.isNaN(anything) === false) && (Number.isFinite(anything) === true));

const isObject = (anything) => (Object.prototype.toString.call(anything) === "[object Object]");

const isArray = (anything) => ((Object.prototype.toString.call(anything) === "[object Array]") && (Array.isArray(anything) === true));

const isFunction = (anything) => (Object.prototype.toString.call(anything) === "[object Function]");

const getType = (anything) => ((isNull(anything) === true) ? jsType.Null : ((isBoolean(anything) === true) ? jsType.Boolean : ((isString(anything) === true) ? jsType.String : ((isNumeric(anything) === true) ? jsType.Numeric : ((isObject(anything) === true) ? jsType.Object : ((isArray(anything) === true) ? jsType.Array : ((isFunction(anything) === true) ? jsType.Function : Object.prototype.toString.call(anything))))))));

const jsonStringify = (anything, { pretty = false } = {}) => {
    // custom JSON.stringify() function jsonStringifyV3
    const indent = " ".repeat(4);
    let indentLevel = 0;
    const jsonStringifyInner = (anythingInner) => {
        const anythingInnerType = getType(anythingInner);
        if (anythingInnerType === jsType.Null) return "null";
        if (anythingInnerType === jsType.String) return `"${anythingInner}"`;
        if ((anythingInnerType === jsType.Numeric) || (anythingInnerType === jsType.Boolean)) return `${anythingInner}`;
        if (anythingInnerType === jsType.Object) {
            if (Object.keys(anythingInner).length === 0) return "{}";
            indentLevel += 1;
            let result = ((pretty === true) ? (`{\n${indent.repeat(indentLevel)}`) : "{ ");
            Object.entries(anythingInner).forEach(([objectKey, objectValue], objectEntryIndex) => {
                result = `${result}"${objectKey}": ${jsonStringifyInner(objectValue)}`;
                if ((objectEntryIndex + 1) !== Object.keys(anythingInner).length) {
                    result = `${result}${((pretty === true) ? (`,\n${indent.repeat(indentLevel)}`) : ", ")}`;
                }
            });
            indentLevel -= 1;
            result = `${result}${((pretty === true) ? (`\n${indent.repeat(indentLevel)}}`) : " }")}`;
            return result;
        }
        if (anythingInnerType === jsType.Array) {
            if (anythingInner.length === 0) return "[]";
            indentLevel += 1;
            let result = ((pretty === true) ? (`[\n${indent.repeat(indentLevel)}`) : "[");
            anythingInner.forEach((arrayItem, arrayItemIndex) => {
                result = `${result}${jsonStringifyInner(arrayItem)}`;
                if ((arrayItemIndex + 1) !== anythingInner.length) {
                    result = `${result}${((pretty === true) ? (`,\n${indent.repeat(indentLevel)}`) : ", ")}`;
                }
            });
            indentLevel -= 1;
            result = `${result}${((pretty === true) ? (`\n${indent.repeat(indentLevel)}]`) : "]")}`;
            return result;
        }
        if (anythingInnerType === jsType.Function) return "[object Function]";
        return anythingInnerType;
    };
    return jsonStringifyInner(anything);
};

console.log("\n// Optional Chaining Operator (?.) in JavaScript");

const JSON_OBJECT = {
    "foo": {
        "bar": "baz",
    },
    "fruits": ["apple", "mango", "banana"],
    "get_rectangle_area": (rectangleWidth, rectangleLength) => (rectangleWidth * rectangleLength)
};
console.log(`JSON_OBJECT: ${jsonStringify(JSON_OBJECT, { pretty: true })}`);

console.log(`JSON_OBJECT?.foo?.bar: ${JSON_OBJECT?.foo?.bar}`);
// JSON_OBJECT?.foo?.bar: baz

console.log(`JSON_OBJECT?.foo?.baz: ${JSON_OBJECT?.foo?.baz}`);
// JSON_OBJECT?.foo?.baz: undefined

console.log(`JSON_OBJECT?.fruits?.[2]: ${JSON_OBJECT?.fruits?.[2]}`);
// JSON_OBJECT?.fruits?.[2]: banana

console.log(`JSON_OBJECT?.fruits?.[5]: ${JSON_OBJECT?.fruits?.[5]}`);
// JSON_OBJECT?.fruits?.[5]: undefined

console.log(`JSON_OBJECT?.get_rectangle_area?.(7, 5): ${JSON_OBJECT?.get_rectangle_area?.(7, 5)}`);
// JSON_OBJECT?.get_rectangle_area?.(7, 5): 35

console.log(`JSON_OBJECT?.get_circle_area?.(7): ${JSON_OBJECT?.get_circle_area?.(7)}`);
// JSON_OBJECT?.get_circle_area?.(7): undefined

console.log(`JSON_OBJECT?.["foo"]?.["bar"]: ${JSON_OBJECT?.["foo"]?.["bar"]}`);
// JSON_OBJECT?.["foo"]?.["bar"]: baz

console.log(`JSON_OBJECT?.["foo"]?.["baz"]: ${JSON_OBJECT?.["foo"]?.["baz"]}`);
// JSON_OBJECT?.["foo"]?.["baz"]: undefined

console.log(`JSON_OBJECT?.["fruits"]?.[2]: ${JSON_OBJECT?.["fruits"]?.[2]}`);
// JSON_OBJECT?.["fruits"]?.[2]: banana

console.log(`JSON_OBJECT?.["fruits"]?.[5]: ${JSON_OBJECT?.["fruits"]?.[5]}`);
// JSON_OBJECT?.["fruits"]?.[5]: undefined

console.log(`JSON_OBJECT?.["get_rectangle_area"]?.(7, 5): ${JSON_OBJECT?.["get_rectangle_area"]?.(7, 5)}`);
// JSON_OBJECT?.["get_rectangle_area"]?.(7, 5): 35

console.log(`JSON_OBJECT?.["get_circle_area"]?.(7): ${JSON_OBJECT?.["get_circle_area"]?.(7)}`);
// JSON_OBJECT?.["get_circle_area"]?.(7): undefined
